### DISCUSSION TITLE
Problems testing zeptoIP

### DISCUSSION THREAD

**Discussion by ctrilsbach** [tabemann/zeptoforth#68] [Q&A] [answered]
I have installed zeptoIP but am having problems with the code snippets in the BUILDING_AND_USING_ZEPTOIP.

These three lines will not compile.

my-cyw43-net init-cyw32-net      - this I think should be my-cyw43-net init-cyw43-net?

my-cyw43-net cyw43-control@ to my-cyw43-control   - Don't know about these two lines
my-cyw43-net net-interface@ to my-interface

Is there any documentation for the my-cyw43* modules ?

**Comment by tabemann** [comment]
Thanks for catching these - `init-cyw32-net` should be what you mentioned, i.e. `init-cyw43-net`, and I forgot to mention that the code should include `simple-w-cyw43-net import`.

As for documentation, I have not gotten around to writing API docs - the only forms of documentation that currently exist are `BUILDING_AND_USING_ZEPTOIP.md` and the `test/rp2040/pico_w_*` test code. Thanks for reminding me that I should actually write API docs for all this.

**Reply by ctrilsbach** [reply]
I am getting an unable to parse error for simple-w-cyw43-net ?

**Reply by tabemann** [reply]
You are getting a parse error for `simple-w-cyw43-net` because I made a silly mistake there; it should really be `simple-cyw43-net`.

**Reply by ctrilsbach** [reply]
Happens to us all. 
It's running but says my SSID is too long or short. It's seven characters.

**Reply by tabemann** [reply]
@ctrilsbach Could you show me your code? I've been using the WiFi at my house for ages (the SSID here is six characters) without any issue.

**Reply by ctrilsbach** [reply]
This is the code I am trying from your guide.

oo import
pico-w-cyw43-net import
simple-cyw43-net import
cyw43-control import
net import
endpoint-process import

0 constant pio-addr
0 constant sm-index
pio::PIO0 constant pio-instance

<pico-w-cyw43-net> class-size buffer: my-cyw43-net
0 value my-cyw43-control
0 value my-interface

pio-addr sm-index pio-instance <pico-w-cyw43-net> my-cyw43-net init-object
my-cyw43-net init-cyw43-net
my-cyw43-net cyw43-control@ to my-cyw43-control
my-cyw43-net net-interface@ to my-interface

: connect-wpa2-ap { D: ssid D: pass - }
  begin ssid pass my-cyw43-control join-cyw43-wpa2 nip until
  my-cyw43-control disable-all-cyw43-events
  my-cyw43-control clear-cyw43-events
  my-cyw43-net run-net-process
;

s" SKYTFD9I" s" password" connect-wpa2-ap

cr ." Discovering IPv4 address..."
my-interface discover-ipv4-addr
my-interface intf-ipv4-addr@
cr ." IPv4 address: " net-misc::ipv4.
my-interface intf-ipv4-netmask@
cr ." IPv4 netmask: " net-misc::ipv4.
my-interface gateway-ipv4-addr@
cr ." Gateway IPv4 address: " net-misc::ipv4.
my-interface dns-server-ipv4-addr@
cr ." DNS server IPv4 address: " net-misc::ipv4.
my-cyw43-net toggle-pico-w-led

**Reply by tabemann** [reply]
Here is my corrected code:

```
oo import
pico-w-cyw43-net import
simple-cyw43-net import
cyw43-control import
net import
endpoint-process import

0 constant pio-addr
0 constant sm-index
pio::PIO0 constant pio-instance

<pico-w-cyw43-net> class-size buffer: my-cyw43-net
0 value my-cyw43-control
0 value my-interface

pio-addr sm-index pio-instance <pico-w-cyw43-net> my-cyw43-net init-object
my-cyw43-net init-cyw43-net
my-cyw43-net cyw43-control@ to my-cyw43-control
my-cyw43-net net-interface@ to my-interface

: connect-wpa2-ap { D: ssid D: pass -- }
begin ssid pass my-cyw43-control join-cyw43-wpa2 nip until
my-cyw43-control disable-all-cyw43-events
my-cyw43-control clear-cyw43-events
my-cyw43-net run-net-process
;

s" ATT949" s" password" connect-wpa2-ap

cr ." Discovering IPv4 address..."
my-interface discover-ipv4-addr
my-interface intf-ipv4-addr@
cr ." IPv4 address: " net-misc::ipv4.
my-interface intf-ipv4-netmask@
cr ." IPv4 netmask: " net-misc::ipv4.
my-interface gateway-ipv4-addr@
cr ." Gateway IPv4 address: " net-misc::ipv4.
my-interface dns-server-ipv4-addr@
cr ." DNS server IPv4 address: " net-misc::ipv4.
my-cyw43-net toggle-pico-w-led
```

Before this I executed under bash:

```
utils/load_cyw43_fw.sh /dev/ttyACM0 ~/Downloads/43439a0.bin  ~/Downloads/43439A0_clm.bin
```

and then:

```
utils/codeload3.sh -B 115200 -p /dev/ttyACM0 serial extra/rp2040/pico_w_net_all.fs
```

I was confused for a moment why the class `<pico-w-cyw43-net>` was not showing up in the code you pasted until I realized that you had not quoted the code as being code, so GitHub saw it as being tags and thus removed it. After that, though, the real cause of the issue was that my example code incorrectly had ``: connect-wpa2-ap { D: ssid D: pass - }`  rather than the correct ` : connect-wpa2-ap { D: ssid D: pass -- }` so `-` was being treated as the name of a new local variable rather than as a comment marker, and thus was pulling an extra value off the stack. (This is why I ought to actually execute my example code verbatim. I feel rather dumb right now.) I have now corrected this in my example code, and it now works, as shown by:

```
Welcome to zeptoforth
Built for rp2040_big, version 1.1.0.2, on Sun Sep 17 08:40:41 PM CDT 2023
zeptoforth comes with ABSOLUTELY NO WARRANTY: for details type `license'
 ok
oo import  ok
pico-w-cyw43-net import  ok
simple-cyw43-net import  ok
cyw43-control import  ok
net import  ok
endpoint-process import  ok
  ok
0 constant pio-addr  ok
0 constant sm-index  ok
pio::PIO0 constant pio-instance  ok
  ok
<pico-w-cyw43-net> class-size buffer: my-cyw43-net  ok
0 value my-cyw43-control  ok
0 value my-interface  ok
  ok
pio-addr sm-index pio-instance <pico-w-cyw43-net> my-cyw43-net init-object  ok
my-cyw43-net init-cyw43-net  ok
my-cyw43-net cyw43-control@ to my-cyw43-control  ok
my-cyw43-net net-interface@ to my-interface  ok
  ok
: connect-wpa2-ap { D: ssid D: pass -- }  ok
begin ssid pass my-cyw43-control join-cyw43-wpa2 nip until  ok
my-cyw43-control disable-all-cyw43-events  ok
my-cyw43-control clear-cyw43-events  ok
my-cyw43-net run-net-process  ok
;  ok
  ok
s" ATT949" s" 3149626775" connect-wpa2-ap  ok
  ok
cr ." Discovering IPv4 address..." 
Discovering IPv4 address... ok
my-interface discover-ipv4-addr  ok
my-interface intf-ipv4-addr@  ok
cr ." IPv4 address: " net-misc::ipv4. 
IPv4 address: 192.168.1.66 ok
my-interface intf-ipv4-netmask@  ok
cr ." IPv4 netmask: " net-misc::ipv4. 
IPv4 netmask: 255.255.255.0 ok
my-interface gateway-ipv4-addr@  ok
cr ." Gateway IPv4 address: " net-misc::ipv4. 
Gateway IPv4 address: 192.168.1.254 ok
my-interface dns-server-ipv4-addr@  ok
cr ." DNS server IPv4 address: " net-misc::ipv4. 
DNS server IPv4 address: 192.168.1.254 ok
my-cyw43-net toggle-pico-w-led  ok
```

**Reply by ctrilsbach** [reply]
Hi Travis, sorry I didn't realise that Github needed the code to be quoted. For the future how do you do it ?

I am now getting the following

r ." Discovering IPv4 address..." 
Discovering IPv4 address... ok
my-interface discover-ipv4-addr 
Starting DHCP discovery
Sending DHCPDISCOVER
Constructing DHCPDISCOVER
Constructed DHCPDISCOVER packet
Waiting for DHCPOFFER
Receiving DHCP packet
Got DHCP message type: 2 
Sending DHCPREQUEST
Sending DHCPREQUEST
Constructing DHCPREQUEST
Constructed DHCPREQUEST packet
Waiting for DHCPACK
Sending DHCPREQUEST
Constructing DHCPREQUEST
Constructed DHCPREQUEST packet
Waiting for DHCPACK
etc etc

**Reply by tabemann** [reply]
To quote a block of code the easiest way to do it is by putting lines before and after it containing just \``` .

As for the extra stuff you are seeing, those are debugging prints for DHCP. I have disabled them by default in more recent versions of zeptoIP and I would suggest you pull the latest because it contains some significant bug fixes (e.g. previously it was sometimes losing data on connection closure).

**Reply by ctrilsbach** [reply]
Great I will give it a go. Thanks Travis, and for the documentation as well.

**Reply by tabemann** [reply]
Just so you know, I have made a fix to the DHCP renewal mechanism, so you might want to git pull the `devel` branch again.

**Reply by ctrilsbach** [reply]
I was pulling down the wrong branch. I have pulled down devel but what ever I do the rp2040_big_uf2 file is only 194k in size after compilation and obviously doesn't work?
This is what I ran. What am I doing wrong ?
```
make VERSION='full_usb'
```

**Reply by tabemann** [reply]
`VERSION` is just the version number being compiled for, which should be something like `1.1.1-dev`. It does not select which builds are to me built. Rather, just execute `make` and it will make an `rp2040_big` kernel by itself.

**Reply by tabemann** [reply]
I should note that precompiled `full_usb` build will not contain a CYW43549 firmware nor any driver or stack; you will need to load those on your own

One little thing I forgot to note - zeptoIP currently requires a `full_usb` build.

**Reply by ctrilsbach** [reply]
So by default rp2040_big is full_usb ? Not that I want a serial interface version but I thought you had to select which version you want to build.

**Reply by tabemann** [reply]
There are both `full` and `full_usb` images for `rp2040_big`,.

Both kernels are built at the same time when you issue:

```
make clean
make VERSION=<your version number, not platform or build type, here>
make install VERSION=<put the same thing here>
```

The kernels will go in `bin/<whatever you said before>/rp2040` and `bin/<whatever you said before>/rp2040_big` and be named `zeptoforth_kernel-<your version  here>.uf2`.

To build your own image, after building the proper kernel and flashing it onto your RP2040 board, you need to use `utils/codeload3.sh -B 115200 -p <device file> serial src/<rp204 or rp2040_big>/forth/setup_<full or full_usb>.fs` to upload the Forth code for the specified build type and platform (the platform must match the kernel you have just flashed onto the board, and the build type must be either `full` or `full_usb`) to simply upload code onto your board, or `utils/make_uf2_image.sh <version> <either rp2040 or rp2040_big> <device file> <full or full_usb>` to build the desired configuration but also pull a UF2 file off the board that you can use again later.

In the premade images, you will find both `full` and `full_big` images at, say, `bin/<version>/rp2040` and `bin/<version>/rp2040_big` with the filenames `zeptoforth_full-<version>.uf2` and `zeptoforth_full_usb-<version>.uf2`. These files are not found in the git repository itself but only in the `zeptoforth-<version>.tar.gz` files you can download from GitHub. These premade images also contain nothing pertaining to zeptoIP or the CYW43439.

**Reply by ctrilsbach** [reply]
I didn't know about the make install.
I still have the same problem though that the rp2040_big_full_usb.uf2 file it is building is only 194k instead of 1.1M and does not work.

**Reply by tabemann** [reply]
Can you paste the exact commands you issued? Thanks.

**Reply by ctrilsbach** [reply]
```
make clean
make VERSION='test'
make install VERSION='test'
```
There are no errors in the output?
I noticed that the development directory tree is 55M and the Master is 92M. Is something missing?

**Reply by ctrilsbach** [reply]
Looking at it again the size of the file is correct. The problem is that it is not building the full_usb files.

**Reply by tabemann** [reply]
No that looks good to me, but note that make by itself does not build anything but the kernel. As for the repository sizes, that is just a side effect of how I was putting binaries in the git tree until I decided it was too big for some users with slow/unreliable connections.

**Reply by ctrilsbach** [reply]
Yes I understand now. Just the Kernel is small I now realise. 
How do I get it generate the full and full_usb files? I am only getting four files in the bin directory.

**Reply by tabemann** [reply]
To generate the full or full_usb files execute:

```
utils/make_uf2_image.sh <the version you picked> {rp2040, rp2040_big} <your tty device> {full, full_usb}
```

where the things in braces are choices you can make.

The files that will be built will go in `bin/<the version you picked>/{rp2040, rp2040_big}` directory depending on which platform and version you selected.

**Reply by ctrilsbach** [reply]
As this requires a tty device does this actually write the build to the Pico ? If so do you need to have a full_usb build of some sort already running on the Pico so there is a /dev/ttyACM0 available?

**Reply by tabemann** [reply]
For initial building you need a serial console, which would most likely have a tty device such as /dev/ttyUSB0. By the time you have a USB CDC console everything in zeptoforth will have already been built except for the code in the `extra` directory.

**Reply by ctrilsbach** [reply]
Getting a lot further but the make_uf2_image.sh script hangs on 'cannot remove screenlog.0' ?

**Reply by tabemann** [reply]
You can ignore the 'cannot remove screenlog.0' message, and the reason why it is apparently hanging is that it is pulling data off the board to construct `.bin`, and from that, `.uf2` files.

**Reply by ctrilsbach** [reply]
Nearly there.

load_cyw43_fw.sh expects the CYW43439 driver files to have a lower case a in the filenames despite what you put in the command line. Thought I was going insane trying to find out why it could not find the files no matter what path you gave them.

At the end of the script it produces an error cyw43_clm  module already defined.

The utils/codeload3.sh runs without any errors.

Trying your test code again it hangs on finding the ip address.
cr ." Discovering IPv4 address..." 
Discovering IPv4 address... ok
my-interface discover-ipv4-addr

**Reply by tabemann** [reply]
Odd about `load_cyw43_fw.sh` because when I use it I actually use a capital A in one of the filenames.

About `cyw43_clm` already being installed the solution is to execute `restore-state` and try uploading the firmware again.

As for the hanging at `discover-ipv4-addr` that just means that it is busy trying to discover an IPv4 address etc. If you want to confirm that it is doing something though you can execute `true to net-config::dhcp-log?` before this, which will make it dump logs to the USB CDC console for DHCP.

**Reply by ctrilsbach** [reply]
Rebuilt everything from scratch and no errors and a much longer time to install.

Getting these three lines continuously now.

```
Constructing DHCPREQUEST
Constructed DHCPREQUEST packet
Waiting for DHCPACK
```

**Reply by tabemann** [reply]
That means it is doing what it is supposed to but if you get this indefinitely it may mean something about you wifi reception.

**Reply by ctrilsbach** [reply]
Strange the router is very close.

Connected fine to my phones mobile hotspot though! I will look at the router settings.

**Reply by ctrilsbach** [reply]
Cannot get ZeptoIP to connect to the router despite having the PicoW less than a inch away.
The only setting is encryption which is WPA2-PSK(AES) on the router. My phones mobile hotspot is WPA2 personal. 
Need to try with other routers.

**Reply by tabemann** [reply]
When I took my Pico W with me on vacation I had a very hard time getting it to talk to the cabins" WiFi for some reason, where it talks with the Wifi AP at my home just fine. I chalked it up at the time to heavy WiFi use, as the AP was shared between multiple cabins, but maybe that was not the issue.

**Reply by tabemann** [reply]
You could try adding

```
cyw43-consts::PM_AGGRESSIVE my-cyw43-control cyw43-power-management
```

or like before you attempt to join an AP.  The power management constants are in `cyw43-consts` and have the following values:

```
  \ Custom. officially unsupported mode. Use at your own risk.
  \ All power-saving features set to their max at only a marginal decrease in
  \ power consumption as opposed to cyw43-pm-aggressive.
  0 constant PM_SUPER_SAVE

  \ Aggressive power saving mode.
  1 constant PM_AGGRESSIVE

  \ The default mode.
  2 constant PM_POWER_SAVE

  \ Performance is preferred over power consumption but still some power is
  \ conserved as opposed to cyw43-pm-none.
  3 constant PM_PERFORMANCE

  \ Unlike all other PM modes, this lowers the power consumption at all times at
  \ the cost of a much lower throughput.
  4 constant PM_THROUGHPUT_THROTTLING

  \ No power management is configured. This conserves the most power.
  5 constant PM_NONE
```

Note that the default is `PM_POWER_SAVE`.

**Reply by ctrilsbach** [reply]
Doesn't make any difference. I tried aggressive and performance. I am still getting the occasional my-cyw43-net RAM dictionary space is running low (<1K left) message

**Reply by tabemann** [reply]
Okay, I did not understand that your RAM - not your flash - was running low. The solution to this is to simply reboot between times you instantiate `<pico-w-cyw43-net>` as that by itself takes up a substantial amount of RAM each time you allot and initialize it.

**Comment by tabemann** [comment]
About the `cyw43*` documentation, the API functionality that is really meant for public consumption is [here](https://github.com/tabemann/zeptoforth/blob/devel/docs/extra/cyw43.md). Note that it is specifically not in the `master` branch because zeptoIP and the CYW43 driver are not in the `master` branch yet. Same goes for the zeptoIP documentation, which is [here](https://github.com/tabemann/zeptoforth/blob/devel/docs/extra/net.md).

**Reply by tabemann** [reply, marked as answer]
And yes, the `cyw43*` documentation may seem somewhat incomplete, but that is really because I do not entirely understand the CYW43439 myself and thus only documented the portions that  would most likely be useful to the user.
